File: go/iterator.go
package msgpackblob

// Iterator is a cursor over a container's children, supporting flat (each) and
// recursive (tree) modes — mirroring the SQLite extension's msgpack_each and
// msgpack_tree table-valued functions.
type Iterator struct {
	blob      Blob
	base      string
	recursive bool
	rows      []EachRow
	cursor    int
	populated bool
}

// NewIterator creates an iterator over the container at path. When recursive is
// true it walks the whole subtree (tree); otherwise it yields direct children
// (each).
func NewIterator(blob Blob, path string, recursive bool) *Iterator {
	if path == "" {
		path = "$"
	}
	return &Iterator{blob: blob, base: path, recursive: recursive, cursor: -1}
}

func (it *Iterator) populate() {
	if it.populated {
		return
	}
	it.populated = true
	it.rows = nil

	a := it.blob.data
	n := len(a)
	if n == 0 {
		return
	}

	iroot := 0
	if it.base != "$" {
		rc, istart, _ := lookup(a, n, 0, it.base)
		if rc != rcOK {
			return
		}
		iroot = istart
	}

	if it.recursive {
		treeWalk(a, n, iroot, it.base, it.base, 0, &it.rows)
	} else {
		it.rows = eachIter(a, n, iroot, it.base)
	}
}

// Next advances the cursor; returns false past the last row.
func (it *Iterator) Next() bool {
	it.populate()
	it.cursor++
	return it.cursor < len(it.rows)
}

// Current returns the row at the cursor (valid after Next returns true).
func (it *Iterator) Current() EachRow {
	return it.rows[it.cursor]
}

// Reset rewinds the cursor to before the first row.
func (it *Iterator) Reset() { it.cursor = -1 }

// Rows returns all rows as a slice.
func (it *Iterator) Rows() []EachRow {
	it.populate()
	return it.rows
}

File: python/msgpack_blob/iterator.py
"""Iterator — a cursor over container children (flat ``each`` / recursive ``tree``)."""

from __future__ import annotations

from typing import Iterator as _PyIterator
from typing import List, Optional

from . import _decode as D
from . import _iterate as I
from ._iterate import EachRow
from .blob import Blob

__all__ = ["Iterator", "EachRow"]


class Iterator:
    """Iterate over a container's children.

    Supports flat (``each``) and recursive (``tree``) modes, mirroring the
    SQLite extension's ``msgpack_each`` / ``msgpack_tree`` table-valued
    functions. Use it as a Python iterator or via the C++-style
    :meth:`next` / :meth:`current` cursor protocol.
    """

    __slots__ = ("_blob", "_base", "_recursive", "_rows", "_cursor", "_populated")

    def __init__(self, blob: Blob, path: str = "$", recursive: bool = False) -> None:
        self._blob = blob
        self._base = path if path else "$"
        self._recursive = recursive
        self._rows: List[EachRow] = []
        self._cursor = -1
        self._populated = False

    def _populate(self) -> None:
        if self._populated:
            return
        self._populated = True
        self._rows = []
        a = self._blob.data()
        n = len(a)
        if n == 0:
            return

        iroot = 0
        if self._base != "$":
            rc, istart, _ = D.lookup(a, n, 0, self._base)
            if rc != D.RC_OK:
                return
            iroot = istart

        if self._recursive:
            I.tree_walk(a, n, iroot, self._base, self._base, 0, self._rows)
        else:
            self._rows = I.each_iter(a, n, iroot, self._base)

    # ── C++-style cursor protocol ─────────────────────────────────────
    def next(self) -> bool:
        self._populate()
        self._cursor += 1
        return self._cursor < len(self._rows)

    def current(self) -> EachRow:
        return self._rows[self._cursor]

    def reset(self) -> None:
        self._cursor = -1

    # ── Pythonic protocols ────────────────────────────────────────────
    def rows(self) -> List[EachRow]:
        self._populate()
        return list(self._rows)

    def __iter__(self) -> _PyIterator[EachRow]:
        self._populate()
        return iter(list(self._rows))
